src/mem/cache/prefetch/bo.hh

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (c) 2024 Samsung Electronics
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met: redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer;
+ * redistributions in binary form must reproduce the above copyright
+ * notice, this list of conditions and the following disclaimer in the
+ * documentation and/or other materials provided with the distribution;
+ * neither the name of the copyright holders nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ * Describes a HPCA variant of Best-Offset prefetcher.
+ */
+
+#ifndef __MEM_CACHE_PREFETCH_BO_HH__
+#define __MEM_CACHE_PREFETCH_BO_HH__
+
+#include "mem/cache/prefetch/queued.hh"
+#include "mem/packet.hh"
+
+namespace gem5
+{
+
+struct BoPrefetcherParams;
+
+namespace prefetch
+{
+
+class Bo : public Queued
+{
+
+  protected:
+      int roundMax, bestOffset, maxScore, badScore, rrSize, offsetIndex, round;
+      const std::vector<int> offsetList {1, 2, 3, 4, 5, 6, 8, 9, 10, 12, 15,
```

Comment:
If this offsetlist is envisioned to be static and constant during the lifetime of the simulation, we can use `static const std::array<Addr, N>` instead.

For address-related variables like offsets, I'd suggest using gem5's `Addr` type. Note that `Addr` is an unsigned 64-bit integer.

Reply:
>If this offsetlist is envisioned to be static and constant during the lifetime of the simulation, we can use static const std::array<Addr, N> instead.
>
>For address-related variables like offsets, I'd suggest using gem5's Addr type. Note that Addr is an unsigned 64-bit integer.

Yes, you are correct. We should have used `static const std::array` and `Addr` instead of `int`. However, since we, @AbhishekUoR and I, don't have time to re-verify the code after making the changes, we would prefer to pass on this suggestion.

However, if you feel that this is a critical change, please let us know and then we can make the changes and re-run experiments to verify the code.